### PR TITLE
feat: Add shape option to joystick

### PR DESCRIPTION
**Pull request by htshah**
Feature add for issue #104.

### CODE REVIEW

**Comment by htshah** [comment]
Changes done:
- [x] Reverted package-lock.json
- [x] Reverted all unrelated style changes

**Comment by darkworks** [comment]
button
canvas

here is joystick shape , let see if we can add it

**Comment by Nek-** [comment]
Hello guys. Any update on this? Should this work be finish? (it seems that only a rebase is required)

**Comment by htshah** [comment]
@yoannmoinet 
Just a follow up for this PR, is there anything else on which you would like to comment?

**Comment by yoannmoinet** [comment]
I'm off with no computer. 
It will have to wait next week end.
I'll add a reminder.

File: src/collection.js [hunk 1]
```diff
@@ -414,13 +416,31 @@ Collection.prototype.processOnMove = function (evt) {
...
+    default:
+        if (dist > size) {
+            dist = size;
+            pos = u.findCoord(nipple.position, dist, angle);
+        }
+
+        xPosition = pos.x - nipple.position.x;
+        yPosition = pos.y - nipple.position.y;
+        break;
+    }
```
[the remark this answers is not in the document]
Comment (yoannmoinet) [reply src/collection.js:443]: Although I think this process should be simplified.
Something like:

```javascript
/* utils.js */
export const clamp = (pos, min, max) => ({
    x: Math.min(Math.max(pos.x, min), max),
    y: Math.min(Math.max(pos.y, min), max)
});

/* collection.js */

// Clamp the position.
if (nipple.options.shape === 'circle') {
    dist = Math.min(dist, size);
    pos = u.findCoord(nipple.position, dist, angle);
} else {
    // Clamp to a square
    pos = u.clamp(nipple.position, -size, size);
    dist = u.distance(pos, nipple.position);
}

var xPosition = pos.x - nipple.position.x;
var yPosition = pos.y - nipple.position.y;

```

What do you think?

**Review by yoannmoinet — CHANGES_REQUESTED** [review]
Thank you for taking the time to implement this.
I have a few suggestions though.
Also, you're missing the types, you need to update [./types/index.d.ts](https://github.com/yoannmoinet/nipplejs/blob/master/types/index.d.ts) to add the new option.

File: src/nipple.js [hunk 2]
```diff
@@ -131,8 +133,11 @@ Nipple.prototype.stylize = function () {
     };
 
     u.extend(styles.el, transitStyle);
-    u.extend(styles.back, borderStyle);
-    u.extend(styles.front, borderStyle);
+    u.extend(
+        styles.back,
+        this.options.shape === 'circle' ? circleBorderStyle : squareBorderStyle
+    );
```
[the remark this answers is not in the document]
Comment (htshah) [reply src/nipple.js]: Made the necessary changes for this suggestion!

File: src/collection.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (htshah) [reply src/collection.js]: @yoannmoinet 
Thank you for this great suggestion! I've implemented it but had to make some minor changes to the suggested `clamp()` function. Here's the updated function:
```js
/* utils.js */
// Clamp position within the range
export const clamp = (pos, nipplePos, size) => ({
    //                          left-clamping        right-clamping
    x: Math.min(Math.max(pos.x, nipplePos.x - size), nipplePos.x + size),
    //                          top-clamping         bottom-clamping
    y: Math.min(Math.max(pos.y, nipplePos.y - size), nipplePos.y + size)
});
```
Also, I've replaced switch statements with if-else statements as you requested.
Do let me know if there are any further changes required!

Thanks!

**Review by yoannmoinet — APPROVED** [review]
Well done, way simpler now.
LGTM.

**Comment by yoannmoinet** [comment]
There are conflicts. 
Once they are fixed I'll merge the PR.

**Comment by yoannmoinet** [comment]
NVM, I'll squash and merge.

**Comment by yoannmoinet** [comment]
Published in `v0.8.6`

**Comment by htshah** [comment]
Thank you soo much for accepting this PR. It is my first ever open source contribution.😁

**Comment by yoannmoinet** [comment]
No problem @htshah that was a good contribution.
Well done, and thank YOU!

**Pull request merged**